main/src/ca/uwaterloo/flix/language/phase/Namer.scala

```diff
@@ -363,19 +381,30 @@ object Namer extends Phase[WeededAst.Program, NamedAst.Root] {
   }
 
   /**
-    * Performs naming on the given definition declaration `decl0` under the given environments `env0`, `uenv0`, and `tenv0`.
+    * Checks the type `tpe` of the signature to ensure that it contains the class type parameter `classTparam`.
     */
-  private def visitDef(decl0: WeededAst.Declaration.Def, uenv0: UseEnv, tenv0: Map[String, Type.Var], ns0: Name.NName)(implicit flix: Flix): Validation[NamedAst.Def, NameError] = decl0 match {
+  private def checkSigType(classTparam: NamedAst.TypeParam, tpe: WeededAst.Type, loc: SourceLocation): Validation[Unit, NameError] = {
+    if (freeVars(tpe).exists(_.name == classTparam.name.name)) {
```

Comment:
What does this check do? Are you requiring that the type variable occurs in the signature somewhere?

Reply:
Yes, this is required for two reasons:
1. I don't think it makes sense to have a signature without the type variable. Shouldn't that just be a regular function outside the class?
1. It's necessary to infer which instance a signature is referring to when it is used. This is why it's a "hard" error rather than a redundancy error.

Reply:
Makes sense.